services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/Application.java

```diff
@@ -40,128 +36,52 @@
-        this.deviceConnectionService = Objects.requireNonNull(deviceConnectionService);
-    }
-
-    /**
-     * Sets the authentication service implementation this server is based on.
-     *
-     * @param authenticationService The authenticationService to set.
-     * @throws NullPointerException if service is {@code null}.
-     */
-    @Autowired
-    public final void setAuthenticationService(final AuthenticationService authenticationService) {
-        this.authenticationService = Objects.requireNonNull(authenticationService);
+    public void setHealthCheckProviders(final List<HealthCheckProvider> healthCheckProviders) {
+        this.healthCheckProviders = healthCheckProviders;
     }
 
     @Override
-    protected final Future<Void> deployRequiredVerticles(final int maxInstances) {
-
-        final Future<Void> result = Future.future();
-        CompositeFuture.all(
-                deployAuthenticationService(), // we only need 1 authentication service
-                deployTenantService(),
-                deployRegistrationService(),
-                deployCredentialsService(),
-                deployDeviceConnectionService()).setHandler(ar -> {
-            if (ar.succeeded()) {
-                result.complete();
-            } else {
-                result.fail(ar.cause());
-            }
-        });
-        return result;
-    }
+    protected final Future<?> deployVerticles() {
 
-    private Future<String> deployCredentialsService() {
-        final Future<String> result = Future.future();
-        log.info("Starting credentials service {}", credentialsService);
-        getVertx().deployVerticle(credentialsService, result);
-        return result;
-    }
+        return super.deployVerticles().compose(ok -> {
 
-    private Future<String> deployAuthenticationService() {
-        final Future<String> result = Future.future();
-        if (!Verticle.class.isInstance(authenticationService)) {
-            result.fail("authentication service is not a verticle");
-        } else {
-            log.info("Starting authentication service {}", authenticationService);
-            getVertx().deployVerticle((Verticle) authenticationService, result);
-        }
-        return result;
-    }
+            @SuppressWarnings("rawtypes")
+            final List<Future> futures = new LinkedList<>();
 
-    private Future<String> deployRegistrationService() {
-        final Future<String> result = Future.future();
-        log.info("Starting registration service {}", registrationService);
-        getVertx().deployVerticle(registrationService, result);
-        return result;
-    }
+            for (final Verticle verticle : this.verticles) {
+                log.info("Deploying: {}", verticle);
+                final Future<String> result = Future.future();
+                getVertx().deployVerticle(verticle, result);
+            }
```

Comment:
A custom device registry component using this `Application` class as a blueprint might find itself in a situation where not *all* verticle beans in the componentScan should be deployed here.
Deploying the service instances one-by-one again in this method doesn't look so straightforward, the service interfaces (like `RegistrationService`) not being `Verticle` instances anymore (instead `BaseRegistrationService` needs to be used) and the service factory verticles (e.g. from `deviceRegistryAmqpServerFactory`) not being deployed via the base class anymore (the `Application` class here doesn't extend `AbstractApplication` with its `deployServiceVerticles` method).

I would find it clearer, if the `Application` class here would also extend `AbstractApplication` and not just deploy *all* verticle beans.

Reply:
Extending from `AbstractApplication` does require at least one `ObjectFactory` to be present. Otherwise it will fail.

However the application at hand doesn't use anything of that. The question is, does this implementation service as a blueprint for writing Spring Boot applications? Or does it serve as an example device registry.

I think the previous model of the `AbstractApplication` was extremely hard to understand, and I hear quite some complaints about that.

Now I think: a) we have what is required for this specific task, and not over-engineer b) people can extend this in a spring-native way, as they want to use spring.

Right now it is simple to understand how this is wired up. Before it was not. And if people want more complexity, they can always introduce that. But I don't think we should create an uber-Application startup framework.